src/common/tags.c

```diff
@@ -922,27 +1121,154 @@ uint32_t dt_tag_get_with_usage(const gchar *keyword, GList **result)
+  }
+  gchar **tokens = g_strsplit(synonyms_entry, ",", 0);
+  if(tokens)
+  {
+    gchar **entry = tokens;
+    while (*entry)
+    {
+      // remove leading and trailing spaces
+      char *e = *entry + strlen(*entry) - 1;
+      while(*e == ' ' && e > *entry)
+      {
+        *e = '\0';
+        e--;
+      }
+      e = *entry;
+      while(*e == ' ' && *e != '\0') e++;
+      if(*e)
+      {
+        synonyms = dt_util_dstrcat(synonyms, "%s, ", e);
+      }
+      entry++;
+    }
+    if (synonyms)
+      synonyms[strlen(synonyms) - 2] = '\0';
+  }
+  g_strfreev(tokens);
+  return synonyms;
+}
+
+gchar *dt_tag_get_synonyms(gint tagid)
+{
+  sqlite3_stmt *stmt;
+  gchar *synonyms = NULL;
+
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                              "SELECT description FROM data.tags WHERE id = ?1 ",
+                              -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, tagid);
+
+  if (sqlite3_step(stmt) == SQLITE_ROW)
+  {
+    synonyms = g_strdup((char *)sqlite3_column_text(stmt, 0));
+  }
+  sqlite3_finalize(stmt);
+  return synonyms;
+}
+
+void dt_tag_set_synonyms(gint tagid, gchar *synonyms_entry)
+{
+  if (!synonyms_entry) return;
+  char *synonyms = dt_cleanup_synonyms(synonyms_entry);
+
+  sqlite3_stmt *stmt;
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                              "UPDATE data.tags SET description = ?2 WHERE id = ?1 ",
+                              -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, tagid);
+  DT_DEBUG_SQLITE3_BIND_TEXT(stmt, 2, synonyms, -1, SQLITE_TRANSIENT);
+  sqlite3_step(stmt);
+  sqlite3_finalize(stmt);
+  g_free(synonyms);
+}
+
+gint dt_tag_get_flags(gint tagid)
+{
+  sqlite3_stmt *stmt;
+
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                              "SELECT flags FROM data.tags WHERE id = ?1 ",
+                              -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, tagid);
+
+  gint flags = 0;
+  if (sqlite3_step(stmt) == SQLITE_ROW)
+  {
+    flags = sqlite3_column_int(stmt, 0);
+  }
+  sqlite3_finalize(stmt);
+  return flags;
+}
+
+void dt_tag_set_flags(gint tagid, gint flags)
+{
+  sqlite3_stmt *stmt;
+
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                              "UPDATE data.tags SET flags = ?2 WHERE id = ?1 ",
+                              -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, tagid);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 2, flags);
+  sqlite3_step(stmt);
+  sqlite3_finalize(stmt);
+}
+
+void dt_tag_add_synonym(gint tagid, gchar *synonym)
+{
+  char *synonyms = dt_tag_get_synonyms(tagid);
+  if (synonyms)
+  {
+    synonyms = dt_util_dstrcat(synonyms, ", %s", synonym);
+  }
+  else
+  {
+    synonyms = g_strdup(synonym);
+  }
+  sqlite3_stmt *stmt;
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                              "UPDATE data.tags SET description = ?2 WHERE id = ?1 ",
```

Comment:
description = ?2
and ?2 is synonyms
and synonyms can be "first, second".

That cannot work or I'm missing something.

Reply:
Yes, synonyms = description field = "first, second, third", or more.
I don't understand your doubt.

![image](https://user-images.githubusercontent.com/23012047/62965442-6fa3bd00-bddb-11e9-8975-993e64937163.png)